### PR TITLE
Update Geolocation permission bubble

### PR DESCRIPTION
**Pull request by simonhong**
fix https://github.com/brave/brave-browser/issues/16897

Added more information to users for getting precise location.
With passed `enableHighAccuracy`, permission bubble could give more precise info.

When page requests current position **_w/o_** `highAccuracy`, below bubble should shown.
[**Type A**]
image

When page requests current position **_with_** `highAccuracy`, we have two types of bubble.
When os location service is **_enabled_**, below bubble is shown
[**Type B**]
image
When os location service is **_disabled_**, this is shown.
[**Type C**]
image

Resolves

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

`GeolocationPermissionRequestBrowserTest.SetEnabledHighAccuracyTest`
`LocalizationUtilTest.GetLocalizedResourceUTF16StringWithPlaceholders`

_NOTE_: geolocation permission bubble is not changed on Linux

1. Launch browser and open inspector with any tabs
2. Run `navigator.geolocation.getCurrentPosition(() => {}, () => {}, { enableHighAccuracy : false });`
3. check **Type A** bubble is shown
4. Dismiss bubble and reset permission state if needed
5. Disable OS location service
6. Run `navigator.geolocation.getCurrentPosition(() => {}, () => {}, { enableHighAccuracy : true });`
7. Chek **Type B** bubble is shown
8. Dismiss bubble and reset permission state if needed
9. Enable OS location service (On macOS, brave should be enabled also)
10. Run `navigator.geolocation.getCurrentPosition(() => {}, () => {}, { enableHighAccuracy : true });`
11. Check **Type C** bubble is shown

### CODE REVIEW

File: chromium_src/chrome/browser/ui/views/permissions/permission_prompt_bubble_base_view.cc [hunk 1]
```diff
@@ -42,10 +48,24 @@
...
 #include "brave/browser/widevine/widevine_permission_request.h"
 #endif
 
+constexpr char kLearnMoreURL[] =
+#if BUILDFLAG(IS_WIN)
+    "https://support.microsoft.com/en-us/windows/"
+    "windows-location-service-and-privacy-3a8eee0a-5b0b-dc07-eede-"
+    "2a5ca1c49088";
+#else
+    // TODO(simonhong): Use different url for other platforms.
```
Comment (simonhong) [inline chromium_src/chrome/browser/ui/views/permissions/permission_prompt_bubble_base_view.cc:65]: @rebron We need different learn more link for macOS or linux.

Comment (simonhong) [reply chromium_src/chrome/browser/ui/views/permissions/permission_prompt_bubble_base_view.cc]: Used https://support.apple.com/guide/mac-help/allow-apps-to-detect-the-location-of-your-mac-mh35873/mac for macOS
Used https://help.ubuntu.com/stable/ubuntu-help/privacy-location.html on linux. it's ubuntu specific link so not sure it's suitable for this learn more link.

**Comment by simonhong** [comment]
If permission bubble knows about whether web site requested geolocation with high accuracy or not,
it would be helpful to give more precise info(warning) to users about geolocation permission request from web site,

**Comment by simonhong** [comment]
`PermissionPromptBubbleOneOriginViewTest.AccessibleTitleDoesNotMentionTooManyPermissions` is failed in PR builder but it's ok on local build. checking.

File: browser/ui/geolocation/geolocation_utils_mac.mm [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  // Service is off globally.
+  if (![CLLocationManager locationServicesEnabled]) {
+    return false;
+  }
+
+  CLLocationManager* manager = [[CLLocationManager alloc] init];
+  if (@available(macOS 11.0, *)) {
+    auto status = [manager authorizationStatus];
+    LOG(ERROR) << __func__ << " ##### " << (int)status;
+    if (status == kCLAuthorizationStatusAuthorized) {
```
Comment (simonhong) [inline browser/ui/geolocation/geolocation_utils_mac.mm:24]: With local build, manager gives always not determined state.
Need to check this with signed build.

File: browser/brave_content_browser_client.cc [hunk 3]
```diff
@@ -564,6 +565,18 @@ void BraveContentBrowserClient::
...
+  associated_registry.AddInterface<
+      geolocation::mojom::BraveGeolocationPermission>(base::BindRepeating(
+      [](content::RenderFrameHost* render_frame_host,
+         mojo::PendingAssociatedReceiver<
+             geolocation::mojom::BraveGeolocationPermission> receiver) {
+        BraveGeolocationPermissionTabHelper::BindBraveGeolocationPermission(
+            std::move(receiver), render_frame_host);
+      },
+      &render_frame_host));
+#endif  // BUILDFLAG(ENABLE_WIDEVINE)
```
Comment (bsclifton) [inline browser/brave_content_browser_client.cc:578]:
```suggestion
#endif  // !BUILDFLAG(IS_ANDROID)
```

File: browser/ui/geolocation/geolocation_utils_mac.mm [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+bool IsSystemLocationSettingEnabled() {
+  // Service is off globally.
+  if (![CLLocationManager locationServicesEnabled]) {
+    return false;
+  }
+
+  CLLocationManager* manager = [[CLLocationManager alloc] init];
+  if (@available(macOS 11.0, *)) {
+    auto status = [manager authorizationStatus];
+    if (status == kCLAuthorizationStatusAuthorized) {
```
Comment (bsclifton) [inline browser/ui/geolocation/geolocation_utils_mac.mm:21]: I think it's good to link to the docs for this; I went to look at https://developer.apple.com/documentation/corelocation/clauthorizationstatus/kclauthorizationstatusauthorized

It shows as working on macOS 10.6 and newer (Lion I think); 11.0 would be Big Sur. Is that right?

Docs show `kCLAuthorizationStatusAuthorized` as deprecated for other platforms (macOS it's OK). What does the prompt look like for enable location services on macOS? Is it only yes/no? Or does macOS also support like iOS has `Always` or `When in use`?

Comment (simonhong) [reply browser/ui/geolocation/geolocation_utils_mac.mm:21]: This code doesn't launch a prompt.  This code just check current state.
Maybe application launch that prompt. but not sure we want to do it?

**Review by bsclifton — COMMENTED** [review]
Reached out via DM about with a few suggestions for `chromium_src/chrome/browser/ui/views/permissions/permission_prompt_bubble_base_view.cc` 😄 

Will stay tuned for updates - and can rereview once addressed

File: browser/ui/geolocation/geolocation_utils_mac.mm (hunk 4 above)
Comment (simonhong) [reply browser/ui/geolocation/geolocation_utils_mac.mm]: All addressed. PTAL again :)

File: chromium_src/chrome/browser/ui/views/permissions/permission_prompt_bubble_base_view.cc [hunk 5]
```diff
@@ -176,6 +198,169 @@ class PermissionLifetimeCombobox : public views::Combobox,
...
+
+  // Learn more is last part.
+  const int learn_more_offset = placeholders.size() - 2;
+  contents_label->AddStyleRange(
+      gfx::Range(offsets[learn_more_offset], offsets[learn_more_offset + 1]),
+      learn_more_style);
+  constexpr int kFixedLabelWidth = 290;
+  contents_label->SizeToFit(kFixedLabelWidth);
+  return contents_label;
+}
```
[the remark this answers is not in the document]
Comment (bsclifton) [reply chromium_src/chrome/browser/ui/views/permissions/permission_prompt_bubble_base_view.cc:263]: Some of it was cleaned up with `brave_l10n::GetStringFUTF16WithPlaceHolders`. Is there something specific you can recommend for bolding certain placeholders in an l10n friendly way?

[the remark this answers is not in the document]
Comment (simonhong) [reply chromium_src/chrome/browser/ui/views/permissions/permission_prompt_bubble_base_view.cc:263]: Hmm, not sure what I can do more here.

File: app/brave_generated_resources.grd [hunk 6]
```diff
@@ -557,6 +557,18 @@ Or change later at <ph name="SETTINGS_EXTENIONS_LINK">$2<ex>brave://settings/ext
...
+          This site has requested your <ph name="PART_ONE">$1</ph>general location<ph name="PART_ONE_END">$2</ph>. <ph name="PART_TWO">$3</ph>Learn more<ph name="PART_TWO_END">$4</ph>
+        </message>
+        <message name="IDS_GEOLOCATION_PERMISSION_BUBBLE_HIGH_ACCURACY_WITH_LOCATION_SERVICE_LABEL" desc="The text for geolocation permission bubble with high accuracy">
+          This site has requested your <ph name="PART_ONE">$1</ph>precise location<ph name="PART_ONE_END">$2</ph>. <ph name="PART_TWO">$3</ph>Learn more<ph name="PART_TWO_END">$4</ph>
+        </message>
+        <message name="IDS_GEOLOCATION_PERMISSION_BUBBLE_HIGH_ACCURACY_WITHOUT_LOCATION_SERVICE_LABEL" desc="The text for geolocation permission bubble with high accuracy">
+          This site has requested your <ph name="PART_ONE">$1</ph>precise location<ph name="PART_ONE_END">$2</ph>. Brave will only able to provide <ph name="PART_TWO">$3</ph>general location<ph name="PART_TWO_END">$4</ph> data due to <ph name="PART_THREE">$5</ph>Location Services<ph name="PART_THREE_END">$6</ph> being disabled. <ph name="PART_FOUR">$7</ph>Learn more<ph name="PART_FOUR_END">$8</ph>
+        </message>
```
[the remark this answers is not in the document]
Comment (simonhong) [reply app/brave_generated_resources.grd:570]: Added `Each part will be used to get offset in the string.` to the desc.

File: components/brave_geolocation_permission/common/BUILD.gn [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+# You can obtain one at https://mozilla.org/MPL/2.0/.
+
+import("//mojo/public/tools/bindings/mojom.gni")
+
+mojom("brave_geolocation_permission") {
+  sources = [ "brave_geolocation_permission.mojom" ]
+
+  export_class_attribute_blink = "CORE_EXPORT"
+  export_define_blink = "BLINK_CORE_IMPLEMENTATION=1"
+  export_header_blink = "third_party/blink/renderer/core/core_export.h"
```
Comment (goodov) [inline components/brave_geolocation_permission/common/BUILD.gn:13]: please remove these export variables and add `brave_geolocation_permission_blink` directly into blink core via `brave_blink_renderer_core_deps`.

File: chromium_src/chrome/browser/ui/views/permissions/permission_prompt_bubble_base_view.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (simonhong) [reply chromium_src/chrome/browser/ui/views/permissions/permission_prompt_bubble_base_view.cc:264]: Addressed via 8ec14a55d1450411fc3a904f8035b7a5487694e9

File: components/brave_geolocation_permission/common/BUILD.gn (hunk 7 above)
Comment (simonhong) [reply components/brave_geolocation_permission/common/BUILD.gn:13]: Addressed via 17ba9845d06515106ea3221f19c3192e4da9e1bd but it seems above defines are needed. 
Got build failure w/o them.

**Review by goodov — APPROVED** [review]

File: chromium_src/chrome/browser/ui/views/permissions/permission_prompt_bubble_base_view.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (bsclifton) [reply chromium_src/chrome/browser/ui/views/permissions/permission_prompt_bubble_base_view.cc:264]: Nicely done! This is clean 😄

**Review by bsclifton — APPROVED** [review]
Tried the different states - each worked great 😄👍 

Nice cleanup on the placeholders too - thanks @goodov for the help

**Comment by simonhong** [comment]
It's finally fixed. thanks all for great review!

**Comment by rishiraj88** [comment]
Great thanks to you, @simonhong , for taking this humongous activity up. It has been as huge to fix as useful it is to end-users.

File: components/resources/brave_components_strings.grd [hunk 8]
```diff
@@ -797,6 +797,10 @@
         You're viewing a WebTorrent page
       </message>
 
+      <message name="IDS_TEST_STRING_FOR_PLACEHOLDERS" desc="Only for test purpose. Don't need to translate">
```
Comment (mkarolin) [inline components/resources/brave_components_strings.grd:800]: @simonhong, you can use `translateable="false"` attribute in the future if the string doesn't need to be translated.

Comment (simonhong) [reply components/resources/brave_components_strings.grd:800]: Got it, thanks!

**Pull request merged**